central/resourcecollection/datastore/datastore_impl.go

```diff
@@ -228,9 +348,32 @@ func (ds *datastoreImpl) DryRunAddCollection(ctx context.Context, collection *st
 		return err
 	}
 
+	if collection == nil {
+		return errors.New("passed collection must not be nil")
+	}
+
 	return ds.dryRunAddCollectionInGraph(collection)
 }
 
+func (ds *datastoreImpl) UpdateCollection(ctx context.Context, collection *storage.ResourceCollection) error {
+	if collection == nil {
+		return errors.New("passed collection must not be nil")
+	}
+
+	// update graph first to detect cycles
+	edgeUpdates, err := ds.updateCollectionInGraph(collection)
+	if err != nil {
+		return err
+	}
+
+	err = ds.storage.Upsert(ctx, collection)
+	if err != nil {
+		// if we fail to upsert, try to restore the graph
+		ds.undoEdgeUpdatesInGraphWithLock(edgeUpdates)
```

Comment:
This could error as well. Capture error and return. This makes another case for copy operation because once the graph is bad state there is no way to recover other than Central restart.
```
a. Obtain lock
b. Create a graph copy.
c. Do all the changes on copy.
d. If successful, `ds.graph = clonedGraph`. 
e. Release lock.
```

Note this workflow is slightly different than dry run workflow when it comes to locking.